### PR TITLE
CAMEL-11267 Add SpanDecorator for the 'rest' component

### PR DESCRIPTION
**Pull request by objectiser**

### CODE REVIEW

**Comment by objectiser** [comment]
This PR fixes an omission from the original opentracing contribution, a span decorator for the 'rest' component. So if possible can this please be added to the 2.19 branch as well.

It enables the traces for REST endpoints to be more clear - using the URI template as the operation name and extracting path parameters as tags, as shown in this trace of the camel-example-restlet-jdbc example.

jaeger-camel-rest-jdbc

**Comment by davsclaus** [comment]
I think we use curly brackets, eg would it not be `/foo/{personId}`, and thus the logic to detect that should be using `{` instead of `(`. 

For example in the rest producer
https://github.com/apache/camel/blob/master/camel-core/src/main/java/org/apache/camel/component/rest/RestProducer.java#L158

**Comment by objectiser** [comment]
@davsclaus I thought that aswell - but in the camel-example-restlet-jdbc it resulted in ( ). Should I change it to handle both?

**Comment by objectiser** [comment]
Or is this just how the template is displayed from the getFromEndpoint().getEndpointUri() - may be it is changed somewhere?

**Comment by davsclaus** [comment]
That would be strange as the uri ought to preserve those curly brackets.

**Comment by objectiser** [comment]
The attached image above shows the operation name, which is taken from the getFromEndpoint().getEndpointUri().

**Comment by objectiser** [comment]
I could change it to work from the endpoint supplied with the events, rather than the exchange.getFromEndpoint() version - main reason I choose this URI is that it had stripped out the ':' separating the path from uriTemplate.

**Comment by davsclaus** [comment]
Ah okay its restlet that is being special it updates its own uri 
https://github.com/apache/camel/blob/master/components/camel-restlet/src/main/java/org/apache/camel/component/restlet/RestletEndpoint.java#L361

You should likely check for both pairs, as eg other rest components wont use `()` but `{}` pairs

**Comment by objectiser** [comment]
@davsclaus I've updated to use the main endpoint URI and curly braces. However this is still specific to the rest component/dsl. I need to find a restlet component example and potentially refactor to make common - but could do it in this PR or in a separate issue. Let me know which you would prefer.

File: components/camel-opentracing/src/main/java/org/apache/camel/opentracing/decorators/RestSpanDecorator.java [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+                path = uri.substring(index + 1);
+                index = path.indexOf('?');
+                if (index != -1) {
+                    path = path.substring(0, index);
+                }
+                path = path.replaceAll(":", "");
+                try {
+                    path = URLDecoder.decode(path, "UTF-8");
+                } catch (UnsupportedEncodingException e) {
+                    LOG.error("Failed to decode URL path '" + path + "'", e);
```
Comment (davsclaus) [inline components/camel-opentracing/src/main/java/org/apache/camel/opentracing/decorators/RestSpanDecorator.java]: Dont log at ERROR as there is nothing the user can do. Log a DEBUG or something and say in the error message that this exception is ignored.

**Comment by davsclaus** [comment]
I suggest you forget about the restlet component at this point and use another example. Restlet is not so good IMHO. We should likely fix it and not let it change its url. You are welcome to log a JIRA about this.

**Comment by objectiser** [comment]
https://issues.apache.org/jira/browse/CAMEL-11276

**Comment by davsclaus** [comment]
Thanks for the PR it has been merged. Do you mind closing this?

**Comment by objectiser** [comment]
Thanks

**Pull request closed without merging**